### PR TITLE
`ProcessResultMatcher` created and used in test

### PR DESCRIPTION
**Pull request by eliasyishak**
Addresses issues:
- https://github.com/flutter/flutter/issues/127135
- https://github.com/flutter/flutter/issues/99767

Creates a matcher class that we can use for `ProcessResult` to check for the exit code.

*List which issues are fixed by this PR. You must list at least one issue.*

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/test/host_cross_arch.shard/cache_test.dart [hunk 1]
```diff
@@ -31,6 +31,12 @@ HostPlatform _identifyMacBinaryArch(String path) {
   final ProcessResult result = processManager.runSync(
     <String>['file', _dartBinary.path],
   );
+  expect(
+      result,
+      ProcessResultMatcher(
+        0,
+        stdoutSubstring: '${_dartBinary.path}: Mach-O 64-bit executable',
+      ));
```
Comment (eliasyishak) [inline packages/flutter_tools/test/host_cross_arch.shard/cache_test.dart:39]: An additional example usage of `ProcessResultMatcher` that also matches on the `exitCode` and a substring to find in `stdout` for a given `ProcessResult`

File: packages/flutter_tools/test/integration.shard/test_utils.dart [hunk 2]
```diff
@@ -124,3 +124,81 @@ abstract final class AppleTestUtils {
...
   }
 }
+
+/// Matcher to be used for [ProcessResult] returned
+/// from a process run
+///
+/// Need to provide [expectedExitCode]
+class ProcessResultMatcher extends Matcher {
+  ProcessResultMatcher(
+    this.expectedExitCode, {
```
Comment (christopherfujino) [inline packages/flutter_tools/test/integration.shard/test_utils.dart:134]: can this be optional and default to 0?

File: packages/flutter_tools/test/integration.shard/test_utils.dart [hunk 3]
```diff
@@ -124,3 +124,81 @@ abstract final class AppleTestUtils {
...
+
+  /// Substring to find in the process's stderr
+  final String? stderrSubstring;
+
+  bool _foundStdout = true;
+  bool _foundStderr = true;
+
+  @override
+  Description describe(Description description) {
+    description.add('exitCode: $expectedExitCode');
```
Comment (christopherfujino) [inline packages/flutter_tools/test/integration.shard/test_utils.dart:153]: Can this be of the form `'a process with the exit code $expectedExitCode`

File: packages/flutter_tools/test/integration.shard/test_utils.dart [hunk 4]
```diff
@@ -124,3 +124,81 @@ abstract final class AppleTestUtils {
...
+    if (stdoutSubstring != null) {
+      description.add('\nstdout substring:\n$stdoutSubstring');
```
Comment (christopherfujino) [inline packages/flutter_tools/test/integration.shard/test_utils.dart:155]:
```suggestion
      description.add(' and stdout "$stdoutSubstring"');
```

File: packages/flutter_tools/test/integration.shard/test_utils.dart [hunk 5]
```diff
@@ -124,3 +124,81 @@ abstract final class AppleTestUtils {
...
+    }
+    if (stderrSubstring != null) {
+      description.add('\nstderr substring:\n$stderrSubstring');
```
Comment (christopherfujino) [inline packages/flutter_tools/test/integration.shard/test_utils.dart:158]:
```suggestion
      description.add(' and stderr: "$stderrSubstring"');
```

**Review by christopherfujino — COMMENTED** [review]
some nits about wording, but LGTM thanks!

**Review by christopherfujino — APPROVED** [review]
LGTM

**Pull request merged**